File: scripts/repair_phpmyadmin_routines.js
const fs = require("fs");

const KEYWORD =
  /\b(END\s+IF|END\s+WHILE|END\s+LOOP|END\s+REPEAT|END\s+CASE|END\s+FOR|BEGIN|END|ELSEIF|WHILE|REPEAT|CASE|LOOP|FOR|IF)\b/gi;

const ROUTINE_START =
  /^CREATE\s+(?:DEFINER\s*=\s*(?:`[^`]+`@`[^`]+`|[^\s]+)\s+)?(PROCEDURE|FUNCTION|TRIGGER)\b/im;

const ACCOUNT_REPORTS = `CREATE PROCEDURE \`account_reports\` (IN \`in_query_type\` VARCHAR(40), IN \`in_from_date\` VARCHAR(20), IN \`in_to\` VARCHAR(20), IN \`in_facility_id\` VARCHAR(50), IN \`in_code\` VARCHAR(20), IN \`in_account_description\` VARCHAR(50), IN \`in_subhead\` VARCHAR(50))
BEGIN
  IF in_query_type = 'Trial_Balance' THEN
    SELECT
      a.description,
      gl.account_code,
      gl.account_subhead,
      CASE
        WHEN (SUM(gl.cr) - SUM(gl.dr)) < 0 THEN ROUND(ABS(SUM(gl.cr) - SUM(gl.dr)), 2)
        ELSE 0.00
      END AS debit,
      CASE
        WHEN (SUM(gl.cr) - SUM(gl.dr)) > 0 THEN ROUND(SUM(gl.cr) - SUM(gl.dr), 2)
        ELSE 0.00
      END AS credit
    FROM general_ledger gl
    INNER JOIN \`account\` a
      ON a.head = gl.account_subhead
     AND a.facilityId = in_facility_id
    WHERE gl.facility_id = in_facility_id
    GROUP BY a.description, gl.account_code, gl.account_subhead, a.head
    HAVING debit > 0 OR credit > 0
    ORDER BY a.head;

  ELSEIF in_query_type = 'individual_ledger' THEN
    SELECT
      gl.transaction_date,
      gl.account_code,
      gl.account_description,
      gl.dr AS debit,
      gl.cr AS credit,
      (
        SELECT ROUND(SUM(g2.dr - g2.cr), 2)
        FROM general_ledger g2
        WHERE g2.facility_id = gl.facility_id
          AND g2.account_subhead = gl.account_subhead
          AND (
            g2.transaction_date < gl.transaction_date
            OR (
              g2.transaction_date = gl.transaction_date
              AND g2.transaction_id <= gl.transaction_id
            )
          )
      ) AS balance
    FROM general_ledger gl
    WHERE gl.facility_id = in_facility_id
      AND gl.account_subhead = in_subhead
    ORDER BY gl.transaction_date, gl.transaction_id;

  ELSEIF in_query_type = 'supplier_individual_ledger' THEN
    SELECT
      gl.transaction_date,
      gl.account_code,
      gl.account_description,
      gl.dr AS debit,
      gl.cr AS credit,
      (
        SELECT ROUND(SUM(g2.dr - g2.cr), 2)
        FROM general_ledger g2
        WHERE g2.facility_id = gl.facility_id
          AND g2.account_code = gl.account_code
          AND (
            g2.transaction_date < gl.transaction_date
            OR (
              g2.transaction_date = gl.transaction_date
              AND g2.transaction_id <= gl.transaction_id
            )
          )
      ) AS balance
    FROM general_ledger gl
    WHERE gl.facility_id = in_facility_id
      AND gl.account_code = in_code
    ORDER BY gl.transaction_date, gl.transaction_id;
  END IF;
END$$`;

function stripLiterals(sql) {
  let out = "";
  for (let i = 0; i < sql.length; i++) {
    const ch = sql[i];
    if (ch === "-" && sql[i + 1] === "-") {
      while (i < sql.length && sql[i] !== "\n") i++;
      continue;
    }
    if (ch === "/" && sql[i + 1] === "*") {
      i += 2;
      while (i + 1 < sql.length && !(sql[i] === "*" && sql[i + 1] === "/")) i++;
      i += 1;
      continue;
    }
    if (ch === "'" || ch === '"' || ch === "`") {
      const quote = ch;
      i++;
      out += " ";
      while (i < sql.length) {
        if (sql[i] === "\\" && quote !== "`") {
          i += 2;
          continue;
        }
        if (sql[i] === quote) {
          if (sql[i + 1] === quote) {
            i += 2;
            continue;
          }
          i++;
          break;
        }
        i++;
      }
      i--;
      continue;
    }
    out += ch;
  }
  return out;
}

function closerFor(kind) {
  return {
    BEGIN: "END",
    IF: "END IF",
    WHILE: "END WHILE",
    LOOP: "END LOOP",
    REPEAT: "END REPEAT",
    CASE: "END CASE",
    FOR: "END FOR",
  }[kind];
}

function neededClosers(body) {
  const text = stripLiterals(body);
  const stack = [];
  KEYWORD.lastIndex = 0;
  let match;
  while ((match = KEYWORD.exec(text))) {
    const word = match[0].replace(/\s+/g, " ").toUpperCase();
    const after = text.slice(match.index + match[0].length).trimStart();

    if (word === "IF") {
      if (after.startsWith("(") && !/^\([\s\S]{0,400}?\)\s*THEN\b/i.test(after)) {
        continue;
      }
      stack.push("IF");
    } else if (word === "ELSEIF") {
      continue;
    } else if (word === "BEGIN") {
      stack.push("BEGIN");
    } else if (word === "WHILE") {
      stack.push("WHILE");
    } else if (word === "REPEAT") {
      stack.push("REPEAT");
    } else if (word === "LOOP") {
      stack.push("LOOP");
    } else if (word === "FOR") {
      if (/^EACH\s+ROW\b/i.test(after)) continue;
      stack.push("FOR");
    } else if (word === "CASE") {
      stack.push("CASE");
    } else if (word === "END IF") {
      if (stack[stack.length - 1] === "IF") stack.pop();
    } else if (word === "END WHILE") {
      if (stack[stack.length - 1] === "WHILE") stack.pop();
    } else if (word === "END LOOP") {
      if (stack[stack.length - 1] === "LOOP") stack.pop();
    } else if (word === "END REPEAT") {
      if (stack[stack.length - 1] === "REPEAT") stack.pop();
    } else if (word === "END CASE") {
      if (stack[stack.length - 1] === "CASE") stack.pop();
    } else if (word === "END FOR") {
      if (stack[stack.length - 1] === "FOR") stack.pop();
    } else if (word === "END") {
      const top = stack[stack.length - 1];
      if (top === "CASE" || top === "BEGIN") stack.pop();
    }
  }
  return stack.reverse().map(closerFor);
}

function closeRoutine(sql) {
  let body = sql.replace(/\s+$/, "");
  if (body.endsWith("$$")) body = body.slice(0, -2).replace(/\s+$/, "");
  body = body.replace(/\bJOIN\s+((?:`[^`]+`|\w+))\s+where\b/gi, "JOIN $1 ON");
  const closers = neededClosers(body);
  if (closers.length) {
    body = body.replace(/[;\s]+$/, "");
    body += ";\n" + closers.join(";\n");
  }
  if (!body.endsWith("$$")) body += "$$";
  return body;
}

function stripDefiner(sql) {
  return sql.replace(
    /CREATE\s+DEFINER\s*=\s*(?:`[^`]+`@`[^`]+`|[^\s]+)\s+/gi,
    "CREATE ",
  );
}

function splitRoutines(section) {
  const re =
    /^CREATE\s+(?:DEFINER\s*=\s*(?:`[^`]+`@`[^`]+`|[^\s]+)\s+)?(PROCEDURE|FUNCTION|TRIGGER)\b/gim;
  const starts = [];
  let m;
  while ((m = re.exec(section))) starts.push(m.index);
  if (!starts.length) return [section];
  const parts = [];
  if (starts[0] > 0) parts.push(section.slice(0, starts[0]));
  for (let i = 0; i < starts.length; i++) {
    const end = i + 1 < starts.length ? starts[i + 1] : section.length;
    parts.push(section.slice(starts[i], end));
  }
  return parts;
}

function isRoutine(chunk) {
  return ROUTINE_START.test(chunk.trimStart());
}

function routineName(chunk) {
  const m = chunk.match(
    /CREATE\s+(?:DEFINER\s*=\s*\S+\s+)?PROCEDURE\s+`([^`]+)`/i,
  );
  return m ? m[1] : null;
}

function repairSection(section) {
  const repaired = [];
  for (const chunk of splitRoutines(section)) {
    if (!isRoutine(chunk)) {
      repaired.push(chunk);
      continue;
    }
    if (routineName(chunk) === "account_reports") {
      repaired.push("\n" + ACCOUNT_REPORTS + "\n\n");
      continue;
    }
    let next = closeRoutine(stripDefiner(chunk));
    if (!next.endsWith("\n")) next += "\n";
    if (!next.endsWith("\n\n")) next += "\n";
    repaired.push(next);
  }
  return repaired.join("");
}

function repairDump(text) {
  const starts = [...text.matchAll(/^DELIMITER\s+\$\$\s*$/gm)];
  if (!starts.length) {
    throw new Error("Could not find DELIMITER $$ ... DELIMITER ; routine block");
  }

  let out = "";
  let cursor = 0;
  for (const start of starts) {
    const startIdx = start.index + start[0].length;
    const afterStart = text.slice(startIdx);
    const endMatch = afterStart.match(/^DELIMITER\s+;\s*$/m);
    if (!endMatch) {
      throw new Error("Could not find closing DELIMITER ; after routines");
    }
    const endIdx = startIdx + endMatch.index;
    out += text.slice(cursor, startIdx);
    out += repairSection(text.slice(startIdx, endIdx));
    cursor = endIdx;
  }
  return out + text.slice(cursor);
}

const src = process.argv[2];
const dest = process.argv[3] || src;
const original = fs.readFileSync(src, "utf8");
const fixed = repairDump(original);
fs.writeFileSync(dest, fixed, "utf8");
console.log(`Wrote ${dest} (${fixed.length} chars)`);

File: scripts/repair_phpmyadmin_routines.py
"""Repair phpMyAdmin dumps that replaced ';' with '$$' and truncated routines."""
from __future__ import annotations

import re
import sys
from pathlib import Path

ROUTINE_START = re.compile(
    r"^CREATE\s+(?:DEFINER\s*=\s*(?:`[^`]+`@`[^`]+`|[^\s]+)\s+)?"
    r"(PROCEDURE|FUNCTION|TRIGGER)\b",
    re.IGNORECASE | re.MULTILINE,
)

KEYWORD = re.compile(
    r"\b(END\s+IF|END\s+WHILE|END\s+LOOP|END\s+REPEAT|END\s+CASE|END\s+FOR|"
    r"BEGIN|END|ELSEIF|WHILE|REPEAT|CASE|LOOP|FOR|IF)\b",
    re.IGNORECASE,
)

ACCOUNT_REPORTS = r"""CREATE PROCEDURE `account_reports` (IN `in_query_type` VARCHAR(40), IN `in_from_date` VARCHAR(20), IN `in_to` VARCHAR(20), IN `in_facility_id` VARCHAR(50), IN `in_code` VARCHAR(20), IN `in_account_description` VARCHAR(50), IN `in_subhead` VARCHAR(50))
BEGIN
  IF in_query_type = 'Trial_Balance' THEN
    SELECT
      a.description,
      gl.account_code,
      gl.account_subhead,
      CASE
        WHEN (SUM(gl.cr) - SUM(gl.dr)) < 0 THEN ROUND(ABS(SUM(gl.cr) - SUM(gl.dr)), 2)
        ELSE 0.00
      END AS debit,
      CASE
        WHEN (SUM(gl.cr) - SUM(gl.dr)) > 0 THEN ROUND(SUM(gl.cr) - SUM(gl.dr), 2)
        ELSE 0.00
      END AS credit
    FROM general_ledger gl
    INNER JOIN `account` a
      ON a.head = gl.account_subhead
     AND a.facilityId = in_facility_id
    WHERE gl.facility_id = in_facility_id
    GROUP BY a.description, gl.account_code, gl.account_subhead, a.head
    HAVING debit > 0 OR credit > 0
    ORDER BY a.head;

  ELSEIF in_query_type = 'individual_ledger' THEN
    SELECT
      gl.transaction_date,
      gl.account_code,
      gl.account_description,
      gl.dr AS debit,
      gl.cr AS credit,
      (
        SELECT ROUND(SUM(g2.dr - g2.cr), 2)
        FROM general_ledger g2
        WHERE g2.facility_id = gl.facility_id
          AND g2.account_subhead = gl.account_subhead
          AND (
            g2.transaction_date < gl.transaction_date
            OR (
              g2.transaction_date = gl.transaction_date
              AND g2.transaction_id <= gl.transaction_id
            )
          )
      ) AS balance
    FROM general_ledger gl
    WHERE gl.facility_id = in_facility_id
      AND gl.account_subhead = in_subhead
    ORDER BY gl.transaction_date, gl.transaction_id;

  ELSEIF in_query_type = 'supplier_individual_ledger' THEN
    SELECT
      gl.transaction_date,
      gl.account_code,
      gl.account_description,
      gl.dr AS debit,
      gl.cr AS credit,
      (
        SELECT ROUND(SUM(g2.dr - g2.cr), 2)
        FROM general_ledger g2
        WHERE g2.facility_id = gl.facility_id
          AND g2.account_code = gl.account_code
          AND (
            g2.transaction_date < gl.transaction_date
            OR (
              g2.transaction_date = gl.transaction_date
              AND g2.transaction_id <= gl.transaction_id
            )
          )
      ) AS balance
    FROM general_ledger gl
    WHERE gl.facility_id = in_facility_id
      AND gl.account_code = in_code
    ORDER BY gl.transaction_date, gl.transaction_id;
  END IF;
END$$"""


def strip_literals(sql: str) -> str:
    out = []
    i = 0
    n = len(sql)
    while i < n:
        ch = sql[i]
        if ch == "-" and i + 1 < n and sql[i + 1] == "-":
            while i < n and sql[i] != "\n":
                i += 1
            continue
        if ch == "/" and i + 1 < n and sql[i + 1] == "*":
            i += 2
            while i + 1 < n and not (sql[i] == "*" and sql[i + 1] == "/"):
                i += 1
            i += 2
            continue
        if ch in ("'", '"', "`"):
            quote = ch
            i += 1
            out.append(" ")
            while i < n:
                if sql[i] == "\\" and quote != "`":
                    i += 2
                    continue
                if sql[i] == quote:
                    if i + 1 < n and sql[i + 1] == quote:
                        i += 2
                        continue
                    i += 1
                    break
                i += 1
            continue
        out.append(ch)
        i += 1
    return "".join(out)


def closer_for(kind: str) -> str:
    return {
        "BEGIN": "END",
        "IF": "END IF",
        "WHILE": "END WHILE",
        "LOOP": "END LOOP",
        "REPEAT": "END REPEAT",
        "CASE": "END CASE",
        "FOR": "END FOR",
    }[kind]


def needed_closers(body: str) -> list[str]:
    text = strip_literals(body)
    stack: list[str] = []
    for match in KEYWORD.finditer(text):
        word = re.sub(r"\s+", " ", match.group(0)).upper()
        start = match.start()
        after = text[match.end() : match.end() + 8].lstrip()

        if word == "IF":
            if after.startswith("("):
                continue
            stack.append("IF")
            continue
        if word == "ELSEIF":
            continue
        if word == "BEGIN":
            stack.append("BEGIN")
            continue
        if word == "WHILE":
            stack.append("WHILE")
            continue
        if word == "REPEAT":
            stack.append("REPEAT")
            continue
        if word == "LOOP":
            stack.append("LOOP")
            continue
        if word == "FOR":
            stack.append("FOR")
            continue
        if word == "CASE":
            stack.append("CASE")
            continue
        if word == "END IF":
            if stack and stack[-1] == "IF":
                stack.pop()
            continue
        if word == "END WHILE":
            if stack and stack[-1] == "WHILE":
                stack.pop()
            continue
        if word == "END LOOP":
            if stack and stack[-1] == "LOOP":
                stack.pop()
            continue
        if word == "END REPEAT":
            if stack and stack[-1] == "REPEAT":
                stack.pop()
            continue
        if word == "END CASE":
            if stack and stack[-1] == "CASE":
                stack.pop()
            continue
        if word == "END FOR":
            if stack and stack[-1] == "FOR":
                stack.pop()
            continue
        if word == "END":
            if stack and stack[-1] == "CASE":
                stack.pop()
                continue
            if stack and stack[-1] == "BEGIN":
                stack.pop()
                continue
            continue
    return [closer_for(kind) for kind in reversed(stack)]


def close_routine(sql: str) -> str:
    sql = sql.rstrip()
    terminated = sql.endswith("$$")
    body = sql[:-2].rstrip() if terminated else sql
    body = re.sub(
        r"\bJOIN\s+((?:`[^`]+`|\w+))\s+where\b",
        r"JOIN \1 ON",
        body,
        flags=re.IGNORECASE,
    )
    closers = needed_closers(body)
    if closers:
        body = body.rstrip().rstrip(";")
        body += ";\n" + ";\n".join(closers)
    if not body.endswith("$$"):
        body += "$$"
    return body


def strip_definer(sql: str) -> str:
    return re.sub(
        r"CREATE\s+DEFINER\s*=\s*(?:`[^`]+`@`[^`]+`|[^\s]+)\s+",
        "CREATE ",
        sql,
        flags=re.IGNORECASE,
    )


def split_routines(section: str) -> list[str]:
    starts = [m.start() for m in ROUTINE_START.finditer(section)]
    if not starts:
        return [section]
    parts: list[str] = []
    if starts[0] > 0:
        parts.append(section[: starts[0]])
    for idx, start in enumerate(starts):
        end = starts[idx + 1] if idx + 1 < len(starts) else len(section)
        parts.append(section[start:end])
    return parts


def is_routine(chunk: str) -> bool:
    return bool(ROUTINE_START.match(chunk.lstrip()))


def routine_name(chunk: str) -> str | None:
    match = re.search(
        r"CREATE\s+(?:DEFINER\s*=\s*\S+\s+)?PROCEDURE\s+`([^`]+)`",
        chunk,
        re.IGNORECASE,
    )
    return match.group(1) if match else None


def repair_dump(text: str) -> str:
    delim_start = re.search(r"^DELIMITER\s+\$\$\s*$", text, re.MULTILINE)
    delim_end = re.search(r"^DELIMITER\s+;\s*$", text, re.MULTILINE)
    if not delim_start or not delim_end or delim_end.start() < delim_start.end():
        raise SystemExit("Could not find DELIMITER $$ ... DELIMITER ; routine block")

    before = text[: delim_start.end()]
    section = text[delim_start.end() : delim_end.start()]
    after = text[delim_end.start() :]

    repaired: list[str] = []
    for chunk in split_routines(section):
        if not is_routine(chunk):
            repaired.append(chunk)
            continue
        name = routine_name(chunk)
        if name == "account_reports":
            prefix_nl = "" if not repaired else ""
            repaired.append("\n" + ACCOUNT_REPORTS + "\n\n")
            continue
        chunk = strip_definer(chunk)
        chunk = close_routine(chunk)
        if not chunk.endswith("\n"):
            chunk += "\n"
        if not chunk.endswith("\n\n"):
            chunk += "\n"
        repaired.append(chunk)

    return before + "".join(repaired) + after


def main() -> None:
    src = Path(sys.argv[1])
    dest = Path(sys.argv[2]) if len(sys.argv) > 2 else src
    original = src.read_text(encoding="utf-8", errors="replace")
    fixed = repair_dump(original)
    dest.write_text(fixed, encoding="utf-8", newline="\n")
    print(f"Wrote {dest} ({len(fixed)} chars)")


if __name__ == "__main__":
    main()
